apps/sms/test/unit/sms_test.js

```diff
@@ -343,6 +343,54 @@ suite('SMS App Unit-Test', function() {
         );
         assert.isFalse(checkUncheckAllButton.disabled);
       });
+      
+      test('Read/Unread buttons', function() {
+        var i, readUnreadButton;
+
+        ThreadListUI.startEdit();
+        // Retrieve all inputs
+        var inputs = ThreadListUI.container.getElementsByTagName('input');
+        readUnreadButton = document
+          .querySelector('#threads-read-unread-button');
+
+        // unread button enabled
+        for (i = inputs.length - 1; i >= 0; i--) {
+          inputs[i].checked = true;
+          Threads.get(inputs[i].value).unreadCount = 0;
+          ThreadListUI.selectionHandler.select(inputs[i].value);
+        }
+        ThreadListUI.checkInputs();
+        assert.equal(readUnreadButton.dataset.action, 'mark-as-unread');
+
+        // read button enabled
+        for (i = inputs.length - 1; i >= 0; i--) {
+          inputs[i].checked = true;
+          Threads.get(inputs[i].value).unreadCount = 1;
+          ThreadListUI.selectionHandler.select(inputs[i].value);
+        }
+        ThreadListUI.checkInputs();
+        assert.equal(readUnreadButton.dataset.action, 'mark-as-read');
+
+        // read button enabled
```

Comment:
nit: Same here. Not sure why you want to split the index = 0 case, but you can check index in forEach loop and reset unreadCount to 0 when index is 0.